inspirehep/modules/disambiguation/core/ml/utils.py

```diff
@@ -0,0 +1,369 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2014-2017 CERN.
+#
+# INSPIRE is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction
+
+"""Helpers for author disambiguation.
+
+.. codeauthor:: Gilles Louppe <g.louppe@cern.ch>
+.. codeauthor:: Mateusz Susik <mateusz.susik@cern.ch>
+
+"""
+
+from __future__ import absolute_import, division, print_function
+
+import json
+
+from beard.utils import given_name
+from beard.utils import name_initials
+from beard.utils import normalize_name
+from beard.utils import given_name_initial
+
+
+def load_signatures(signatures_filename, records_filename):
+    """Load signatures from JSON files.
+
+    Parameters
+    ----------
+    :param signatures_filename: string
+        Path to the signatures file. The file should be in json format.
+
+    :param records_filename: string
+        Path to the records file. The file should be in json formaat.
+
+    Returns
+    -------
+    :returns: tuple
+        Signatures and records. Both are lists of dictionaries.
+    """
+
+    signatures = []
+    for line in open(signatures_filename, "r"):
+        signatures.append(json.loads(line))
+    records = json.load(open(records_filename, "r"))
+
+    if isinstance(signatures, list):
+        signatures = {s["signature_uuid"]: s for s in signatures}
+
+    if isinstance(records, list):
+        records = {r["publication_id"]: r for r in records}
+
+    for signature_uuid, signature in signatures.items():
+        signature["publication"] = records[str(signature["publication_id"])]
```

Comment:
This eager join requires a lot of memory: the same `publication` is repeated for each signature, and each `publication` contains a list of all the author names, so a record with lots of signatures explodes quadratically in memory use!

I think that here we should prefer a lazy join.